rfcs/xxxx-collage.md

```diff
@@ -0,0 +1,863 @@
+  the choice of layout, and if two BYOC integrations have a different choice of layout then there's currently no way for
+  them to be used concurrently. All of those passes must either be i) pushed up to global configuration (which could be
+  explored by a search layer outside of TVM), ii) pushed into the BYOC lowering/codegen function (to prepare the
+  sub-graph for further compilation) or iii) moved into the standard Relay optimization passes run before
+  `CollageFuseOps`.
+- **BYOC-specific patterns must match after standard Relay passes**: We currently run `CollageFuseOps` just before
+  `FuseOps`, which means it runs after quite a number of standard Relay passes which themselves may invoke some
+  target-specific rewrites (eg for quantization legalization). However, currently BYOC `partition_for_<toolchain>`
+  functions run before any other Relay passes, and thus the Relay model is seen in it's 'clean' state. We need to
+  figure out how to move `CollageFuseOps` earlier in the pipeline to be before any target-specific rewrites or any 
+  passes which may make BYOC patterns and lowering more difficult, but without interfering with TVM's fusion. This
+  needs more thought.
+- **Higher tuning cost**: Obviously Collage needs to estimate the latency of many more candidate kernels, and each
+  candidate may itself trigger tuning during lowering. For TVM this can require O(thousands) of trials and take O(hours)
+  , so we'll be very dependent on cached tuning logs to amortize this cost between models for the same target.
+  Currently Collage will measure more candidates even if TVM native is the only available target.
+- **Task extraction vs Tuning**: Traditionally TVM has had three phases: i) Task extraction (find the fused sub-graphs
+  to tune), ii) Tuning (find a good schedule for those sub-graphs), and iii) Compilation (re-compile the model, now
+  retrieving schedules for all the anticipated sub-graphs from the cache.) However the Collage 'v2' prototype collapses
+  all these phases. This lets us lazily explore the implied search graph (nodes = partially rewritten models, edges =
+  selected of sub-graph and toolchain as a candidate kernel, cost = estimated sum of kernel costs plus launch penalties)
+  , and thus only pay the cost of tuning candidate kernels which could possibly influence the final partitioning.
+- **No non-local optimization**: Though Collage can explore the choice of sub-graph and toolchain, it cannot explore any
+  choices which require the arguments and/or result of the sub-graph to be rewritten, or the overall `IRModule` to be
+  changed. Thus Collage **cannot** be used to search over:
+    - choice of layout for arguments/results (may require insertion of layout transforms),
+    - choice of memory scope for arguments/results (may require insertion of device copies),
+    - choice of device on which to host the kernel (ditto),
```

Comment:
It would be valuable to consider whether in future there would be a viable way to include device exploration in Collage without major design changes. I think there are a number of potentially very significant opportunites to explore with device tuning, particularly in the mobile space.

Reply:
Yes, same point by Christoph above. I'd like to, but for now trying to draw the line at not requiring search over rewrites, just partitionings.

Reply:
Note to self to expand this a bit. Search over layout, scope and device placement could technically be done since those choices can be encoded with additional transforms/copies/etc on call arguments and results. However there's little use for that since the estimated model latency won't account for any Relay optimization to remove redundant transforms/copies/etc.

I think global optimization for layout, device placement etc should be done before Collage. TQ reminded me yesterday we don't need to solve all the problems in one pass and composable passes can be added before and after Collage.

Reply:
Going to resolve. Expanded on the difficulty (yes can rewrite, but unlikely to help since Collage cost won't account for downstream global optimizations enabled by such.)